### PR TITLE
Feature/Update information of Data Elements to add Calculation mode

### PR DESCRIPTION
**Pull request by nshandra**
### References
Issue: [Update information of Data Elements to specify](https://app.clickup.com/t/3jna3kt)

### Description
- Add information about Calculation mode to programs dataElements.

### CODE REVIEW

File: app/dossierPrograms/dossierPrograms.controllers.js [hunk 1]
```diff
@@ -100,19 +82,129 @@ dossierProgramsModule.controller("dossiersProgramSectionController", [
...
+                            pssDE.calcMode = { type: "default" };
+                        }
+                    });
+                });
+
+                if (hiddenSectionsArray.includes(pss.id)) {
+                    pss.dataElements.forEach(pssDE => {
+                        if (pssDE.calcMode.type !== "programRule") {
+                            pssDE.calcMode = { type: "other" };
+                        }
```
Comment (tokland) [inline app/dossierPrograms/dossierPrograms.controllers.js:119]: [minor] Does it make sense to move this logic to the previous loop?

File: app/dossierPrograms/dossierPrograms.controllers.js [hunk 2]
```diff
@@ -100,19 +82,129 @@ dossierProgramsModule.controller("dossiersProgramSectionController", [
...
+
+                $q.resolve(
+                    dossiersProgramStageCalcModeFactory.get({ programId: $scope.selectedProgram.id }).$promise,
+                    data => ($scope.programRules = data.programRules)
+                ).then(() => {
+                    $q.all(stageSectionPromises).then(function (stages) {
+                        const hiddenSectionsArray = $scope.programRules.flatMap(pr => {
+                            return pr.programRuleActions.flatMap(pra => {
+                                if (pra.programRuleActionType === "HIDESECTION") {
+                                    return pra.programStageSection.id;
```
Comment (tokland) [inline app/dossierPrograms/dossierPrograms.controllers.js]: [nitpick] Since it's a flatMap, it should return an array: `return [pra.programStageSection.id]`

File: app/dossierPrograms/dossierPrograms.controllers.js [hunk 3]
```diff
@@ -100,19 +82,129 @@ dossierProgramsModule.controller("dossiersProgramSectionController", [
...
+        function makeSectionsCalcMode(stage, assignedDEArray, hiddenSectionsArray) {
+            stage.programStageSections.forEach(pss => {
+                pss.dataElements.forEach(pssDE => {
+                    assignedDEArray.forEach(adeArray => {
+                        if (adeArray.ids.includes(pssDE.id)) {
+                            pssDE.calcMode = { type: "programRule", name: adeArray.name };
+                        } else if (!pssDE.calcMode) {
+                            pssDE.calcMode = { type: "default" };
+                        }
+                    });
```
Comment (tokland) [inline app/dossierPrograms/dossierPrograms.controllers.js:114]: This kind of stateful imperative logic looks a bit weird, but I see the project is not using lodash at all, so let's keep it like this.

File: app/dossierPrograms/dossierPrograms.services.js [hunk 4]
```diff
@@ -33,6 +20,13 @@ dossierProgramsModule.factory("dossiersProgramsFactory", [
     },
 ]);
 
+var qryProgramStageSections =
+    dhisUrl +
+    "programStages/:programStageId?fields=id,displayName,repeatable,sortOrder," +
+    "programStageSections[id,displayName,dataElements[id,displayName,displayFormName,displayDescription,valueType," +
+    "optionSetValue,optionSet[name,options[displayName]]]],programStageDataElements[dataElement[id,displayName," +
+    "displayFormName,displayDescription,valueType,optionSetValue,optionSet[name,options[displayName]]]]&paging=false";
+
```
Comment (tokland) [inline app/dossierPrograms/dossierPrograms.services.js]: Same as the other PR, it's hard to follow which fields are being requested.

File: app/dossierPrograms/dossierPrograms.services.js [hunk 5]
```diff
@@ -194,6 +217,9 @@ dossierProgramsModule.factory("dossiersProgramEventReportFactory", [
     },
 ]);
 
+var qryEventCharts =
+    dhisUrl + "eventCharts?filter=program.id\\:eq\\::programId&fields=id,displayName,displayDescription&paging=false";
```
Comment (tokland) [inline app/dossierPrograms/dossierPrograms.services.js]: Probably nothing to worry about here, but as a note (I just discovered it):

> New applications and clients should avoid using the eventCharts and eventReports APIs because they are deprecated. Use the eventVisualizations API instead.

https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-238/visualizations.html#webapi_event_visualization

File: app/dossierPrograms/dossierPrograms.view.html [hunk 6]
```diff
@@ -88,6 +89,17 @@ <h3 class="section" id="{{ section.id }}">{{ section.displayName }}</h3>
                                 <td>{{ dataElement.displayFormName }}</td>
                                 <td>{{ dataElement.displayDescription }}</td>
                                 <td>{{ dataElement.valueType }}</td>
+                                <td>
+                                    <div ng-switch="dataElement.calcMode.type">
+                                        <div ng-switch-when="programRule">
+                                            Program Rule: {{ dataElement.calcMode.name }}
```
Comment (tokland) [inline app/dossierPrograms/dossierPrograms.view.html]: | translate

**Review by tokland — COMMENTED** [review]
All minor comments, code looks good.

File: app/dossierPrograms/dossierPrograms.controllers.js (hunk 1 above)
Comment (nshandra) [reply app/dossierPrograms/dossierPrograms.controllers.js:119]: Yup, no need for two loops.

File: app/dossierPrograms/dossierPrograms.controllers.js (hunk 2 above)
Comment (nshandra) [reply app/dossierPrograms/dossierPrograms.controllers.js]: Done.

File: app/dossierPrograms/dossierPrograms.services.js (hunk 4 above)
Comment (nshandra) [reply app/dossierPrograms/dossierPrograms.services.js]: Fixed following the previous PR proposal 1 style.

File: app/dossierPrograms/dossierPrograms.view.html (hunk 6 above)
Comment (nshandra) [reply app/dossierPrograms/dossierPrograms.view.html]: Done.

File: app/dossierPrograms/dossierPrograms.services.js (hunk 5 above)
Comment (nshandra) [reply app/dossierPrograms/dossierPrograms.services.js]: I will look into this after we finish the first batch of PR's.

File: app/dossierPrograms/dossierPrograms.controllers.js (hunk 3 above)
Comment (nshandra) [reply app/dossierPrograms/dossierPrograms.controllers.js:114]: lodash is among the dependencies and can be used. For example, `dossierPrograms.controllers.js` uses `_.uniq`.

If it helps, the logic is to check if the _sections_ and the _dataElements_ are affected buy _programRules_ of "_HIDESECTION_" or "_ASSIGN_" type.
* "programRuleActionType" == "ASSIGN" ->  Program Rule
* "programRuleActionType" != "ASSIGN" and "HIDESECTION" == true ->  Other
* "programRuleActionType" != "ASSIGN" and "HIDESECTION" == false -> Direct data entry

**Review by tokland — APPROVED** [review]
